internal/sshtunneler/tunneler.go

```diff
@@ -0,0 +1,251 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package sshtunneler
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	gossh "golang.org/x/crypto/ssh"
+
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/pki/ssh"
+	"github.com/juju/juju/state"
+)
+
+const (
+	maxTimeout        = 60 * time.Second
+	reverseTunnelUser = "reverse-tunnel"
+	tokenIssuer       = "sshtunneler"
+	tokenSubject      = "reverse-tunnel"
+	tunnelIDClaim     = "tunnelID"
+)
+
+// State writes requests for tunnels to state.
+type State interface {
+	InsertSSHConnRequest(arg state.SSHConnRequestArg) error
+}
+
+// ControllerInfo fetches the controller's address.
+type ControllerInfo interface {
+	Addresses() network.SpaceAddresses
+}
+
+// SSHDial establishes an SSH connection over the provided connection.
+type SSHDial interface {
+	Dial(conn net.Conn, username string, privateKey gossh.Signer) (*gossh.Client, error)
+}
+
+type tunnelRequest struct {
+	privateKey gossh.Signer
+	dialer     SSHDial
+	recv       chan (net.Conn)
+	cleanup    func()
+}
+
+type TunnelTracker struct {
+	authn      tunnelAuthentication
+	state      State
+	controller ControllerInfo
+	dialer     SSHDial
+	clock      clock.Clock
+
+	mu      sync.Mutex
+	tracker map[string]*tunnelRequest
+}
+
+// TunnelTrackerArgs holds the arguments for creating a new tunnel tracker.
+type TunnelTrackerArgs struct {
+	State          State
+	ControllerInfo ControllerInfo
+	Dialer         SSHDial
+	Clock          clock.Clock
+
+	// SharedSecret is the secret used to sign and validate JWTs.
+	SharedSecret []byte
+	// JWTAlg is the algorithm used to sign JWTs and should match
+	// the strength (number of bytes) of the SharedSecret.
+	JWTAlg jwa.KeyAlgorithm
```

Comment:
We don't need this here or shared secret, in the constructor for NewTunnelTracker we can pass the authentication struct with them set already. It'll read better that way too.

Reply:
actually, do we even need this as an argument.. can't the shared secret be recreated every time we create the tunnel tracker? and the algorithm won't change over time, will it (it's not configurable, so we can use a default value)?

Reply:
Initially I was recreating the shared secret every time we create the tunnel tracker, but Simon rightly suggested if the worker for the tunnel tracker bounces due to a bug we would needlessly recreate the secret. So now we instead just pass in the secret and it will be created outside the dependency engine.